internal/kubernetes/restore/restore.go

```diff
@@ -0,0 +1,65 @@
+package restore
+
+import (
+	"fmt"
+	"os"
+
+	certmanagerv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/jetstack/jsctl/internal/kubernetes/yaml"
+)
+
+// RestoredIssuers contains the issuers and cluster issuers that were extracted
+// from a backup file. Issuers which were unsupported are listed in MissedIssuers.
+type RestoredIssuers struct {
+	CertManagerIssuers        []certmanagerv1.Issuer
+	CertManagerClusterIssuers []certmanagerv1.ClusterIssuer
+
+	MissedIssuers []string
+}
+
+func ExtractOperatorManageableIssuersFromBackupFile(backupFilePath string) (*RestoredIssuers, error) {
+	var restoredIssuers RestoredIssuers
+
+	file, err := os.Open(backupFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open backup file: %w", err)
+	}
+
+	resources, err := yaml.Load(file)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load backup file: %w", err)
+	}
+
+	for _, resource := range resources {
+		if resource.GetAPIVersion() != "cert-manager.io/v1" {
```

Comment:
Do you think that it's worth warning about the non v1 when the backup can only create v1 issuers? I hadn't really expected this command to be used with hand crafted issuer files but I guess we can still instruct them to convert.

I can probably make this command support vei issuers, that'd be best here.

Reply:
>Do you think that it's worth warning about the non v1 when the backup can only create v1 issuers? I hadn't really expected this command to be used with hand crafted issuer files but I guess we can still instruct them to convert.

It's one of the migration scenarios that we have where the issuers file comes from Git config

Reply:
https://docs.google.com/document/d/1AxUss0piE9gb_r31_jBS6mv7zWALDqI0ZIfBUiJj3p0/edit#bookmark=id.k9gx1ok6nznk

Reply:
agree about vei if you think there's time 👍🏼

Reply:
Hmm, so if I find a backed up vei issuer like this I'd need access to the secret to generate a connection to use. I don't think we can really do that without badly overloading the apply command.

```
apiVersion: jetstack.io/v1alpha1
kind: VenafiClusterIssuer
metadata:
  name: application-team-a
spec:
  tpp:
    policyDN: \VED\Policy\Teams\ApplicationTeamA
    url: https://tpp1.example.com

    accessToken:
      - secret:
          name: secret1
          fields: ["username", "password"]
      - tppOAuth:
          authInputType: UsernamePassword
          url: https://tpp1.example.com
```

I feel like this is something that we'd want to spend time on if we actually had this problem. However, I don't think that we're going to have hoards of venafi issuer users migrating via this restore path. Wdyt?

Reply:
In e25d298cbb5994c1074d2cb39b6b3b5af0c70336 I restore the vei issuers from a back up but I don't think we want to do this since we can't complete the loop without the secret really.

Reply:
>  to generate a connection to use

why would you need to do that? you should be able to just translate the VEI issuer config into the operator config.

I assume that you mean the `--experimental-cert-discovery-venafi-connection` flag- this hasn't been made to work with VEI issuers yet, but I think it is not needed for this command- the flag is so as to make it easier for users to provide connections details, but in this case is not needed as we get them from backup. lmk if I am missing something

Reply:
It's just that a VenafiClusterIssuer resource references a secret (rather than containing the credentials in the resource), so in order to recreate the secret in the correct place (wherever the operator is configured to deploy the vei), jsctl would need to go and fetch that right?

Reply:
I can't see why the user would need to delete that secret I think it's fair to assume that it still exists in the cluster?

Reply:
OK, I can have another think about this!

Reply:
7741d1e04392ca432c909d3ffb301d842638eb1e should do what we need.

Reply:
```
...arlieegan/Code/jsctl $ cat /Users/charlieegan/Code/jsctl/internal/kubernetes/restore/fixtures/backup.json
{
  "apiVersion": "v1",
  "kind": "List",
  "items": [
    {
      "apiVersion": "awspca.cert-manager.io/v1beta1",
      "kind": "AWSPCAIssuer",
      "metadata": {
        "creationTimestamp": null,
        "name": "pca-sample",
        "namespace": "jetstack-secure"
      },
      "spec": {
        "arn": "acb",
        "secretRef": {
          "accessKeyIDSelector": {
            "key": ""
          },
          "secretAccessKeySelector": {
            "key": ""
          }
        }
      },
      "status": {
      }
    },
    {
      "apiVersion": "cert-manager.io/v1",
      "kind": "ClusterIssuer",
      "metadata": {
        "creationTimestamp": null,
        "name": "cm-cluster-issuer-sample"
      },
      "spec": {
        "acme": {
          "email": "dummy-email@example.com",
          "preferredChain": "",
          "privateKeySecretRef": {
            "name": "example"
          },
          "server": "https://"
        }
      },
      "status": {
      }
    },
    {
      "apiVersion": "cas-issuer.jetstack.io/v1beta1",
      "kind": "GoogleCASIssuer",
      "metadata": {
        "creationTimestamp": null,
        "name": "googlecasissuer-sample",
        "namespace": "jetstack-secure"
      },
      "spec": {
        "caPoolId": "my-pool",
        "credentials": {
          "key": "example",
          "name": "googlesa"
        },
        "location": "us-east1",
        "project": "example"
      },
      "status": {
      }
    },
    {
      "apiVersion": "cert-manager.io/v1",
      "kind": "Issuer",
      "metadata": {
        "creationTimestamp": null,
        "name": "cm-issuer-sample",
        "namespace": "jetstack-secure"
      },
      "spec": {
        "ca": {
          "secretName": "ca-key-pair"
        }
      },
      "status": {
      }
    },
    {
      "apiVersion": "cert-manager.io/v1",
      "kind": "Certificate",
      "metadata": {
        "creationTimestamp": null,
        "name": "example-com",
        "namespace": "jetstack-secure"
      },
      "spec": {
        "dnsNames": [
          "example.com"
        ],
        "issuerRef": {
          "group": "cert-manager.io",
          "kind": "Issuer",
          "name": "ca-issuer"
        },
        "secretName": "example-com-tls"
      },
      "status": {
      }
    },
    {
      "apiVersion": "jetstack.io/v1alpha1",
      "kind": "VenafiClusterIssuer",
      "metadata": {
        "name": "application-team-a"
      },
      "spec": {
        "tpp": {
          "policyDN": "\\VED\\Policy\\Teams\\ApplicationTeamA",
          "url": "https://tpp1.example.com"
        }
      }
    },
    {
      "apiVersion": "jetstack.io/v1alpha1",
      "kind": "VenafiIssuer",
      "metadata": {
        "name": "application-team-b"
      },
      "spec": {
        "tpp": {
          "policyDN": "\\VED\\Policy\\Teams\\ApplicationTeamA",
          "url": "https://tpp1.example.com"
        }
      }
    },
    {
      "apiVersion": "policy.cert-manager.io/v1alpha1",
      "kind": "CertificateRequestPolicy",
      "metadata": {
        "creationTimestamp": null,
        "name": "test-policy"
      },
      "spec": {
        "allowed": {
          "commonName": {
            "required": true,
            "value": "hello.world"
          }
        },
        "selector": {
          "issuerRef": {
          }
        }
      },
      "status": {
      }
    }
  ]
}
...arlieegan/Code/jsctl $ go run main.go operator installations apply --stdout --experimental-issuers-backup-file=/Users/charlieegan/Code/jsctl/internal/kubernetes/restore/fixtures/backup.json
The following issuers cannot be managed by the operator and must be restored manually: AWSPCAIssuer/pca-sample, GoogleCASIssuer/googlecasissuer-sample
apiVersion: operator.jetstack.io/v1alpha1
kind: Installation
metadata:
  creationTimestamp: null
  name: installation
spec:
  approverPolicy: {}
  certManager:
    controller:
      replicas: 2
    webhook:
      replicas: 2
  issuers:
  - ca:
      secretName: ca-key-pair
    name: cm-issuer-sample
    namespace: jetstack-secure
  - acme:
      email: dummy-email@example.com
      preferredChain: ""
      privateKeySecretRef:
        name: example
      server: https://
    clusterScope: true
    name: cm-cluster-issuer-sample
  - name: application-team-b
    venafiEnhancedIssuer:
      tpp:
        policyDN: \VED\Policy\Teams\ApplicationTeamA
        url: https://tpp1.example.com
  - clusterScope: true
    name: application-team-a
    venafiEnhancedIssuer:
      tpp:
        policyDN: \VED\Policy\Teams\ApplicationTeamA
        url: https://tpp1.example.com
status: {}

```

Reply:
ff884aedd41db1a56e05784855589d553896d546 will show a new message for issuers that must be converted:

```
$ go run main.go operator installations apply --stdout --experimental-issuers-backup-file=/Users/charlieegan/Code/jsctl/internal/kubernetes/restore/fixtures/backup.json
The following issuers cannot be managed by the operator and must be restored manually: AWSPCAIssuer/pca-sample, GoogleCASIssuer/googlecasissuer-sample
The following issuers need to be converted to cert-manager v1 resources: ClusterIssuer/outdated-cm-issuer
This can be done using cmctl convert, see here for more information: https://cert-manager.io/docs/reference/cmctl/#convert
apiVersion: operator.jetstack.io/v1alpha1
kind: Installation
metadata:
  creationTimestamp: null
  name: installation
...
```

These messages are sent to Std error so that they don't appear in the yaml if using --stdout.